Reword ssl_conf_max_frag_len documentation to clarify its necessity

k-stachowiak opened the pull request:
This PR is supposed to fix #2258 by increasing precision of the max fragment length API documentation.

## Needs backports
YES - 2.16 and 2.7

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,18 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  *                 (Client: set maximum fragment length to emit *and*
  *                 negotiate with the server during handshake)
  *
+ * \note           By default the \c mfl_code field ofthe \c mbedtls_ssl_config
```
Typo: `ofthe`

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,18 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  *                 negotiate with the server during handshake)
  *
+ * \note           By default the \c mfl_code field ofthe \c mbedtls_ssl_config
+ *                 structure is equal `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
```
Minor: is equal _to_

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,18 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  *
+ * \note           By default the \c mfl_code field ofthe \c mbedtls_ssl_config
+ *                 structure is equal `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
+ *                 This means the max fragment length extension *will not* be
```
Minor: Spell out 'maximum'

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,18 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
+ * \note           By default the \c mfl_code field ofthe \c mbedtls_ssl_config
+ *                 structure is equal `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
+ *                 This means the max fragment length extension *will not* be
+ *                 used unless set to other value via this function.
```
Typo: "set to _an_other" or "... unless set to a non-zero value via this function."

k-stachowiak commented:
@hanno-arm Thanks for the review. I have addressed your comments.

AndrzejKurek commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,18 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  *
+ * \note           By default the \c mfl_code field of the \c mbedtls_ssl_config
+ *                 structure is equal to `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
+ *                 This means the maximum fragment length extension *will not*
```
I would rather say "This means that"

AndrzejKurek commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,18 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
+ * \note           By default the \c mfl_code field of the \c mbedtls_ssl_config
+ *                 structure is equal to `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
+ *                 This means the maximum fragment length extension *will not*
+ *                 be used unless set to another value via this function.
```
Shouldn't this be "set to a different" instead of "another"? Another value does not imply that it is not equal to the first one.

k-stachowiak commented:
@AndrzejKurek Thanks for the review. I have addressed your comments. If @hanno-arm, who you asked for comments, objects my new changes, I am willing to introduce further changes.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  *                 negotiate with the server during handshake)
  *
+ * \note           By default the \c mfl_code field of the \c mbedtls_ssl_config
+ *                 structure is equal to `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
```
I find the `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE` confusing (does it mean that it's 0, or that it's the result of evaluating the expression). I'd remove the `0 == ` par and just go with the named constant.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
+ *                 structure is equal to `0 == MBEDTLS_SSL_MAX_FRAG_LEN_NONE`.
+ *                 Therefore, the maximum fragment length extension *will not*
+ *                 be used, unless the maximum fragment length has been set to a
+ *                 different value via this function.
```
I think we need to clarify that this is about clients. Unless I'm mistaken, servers will accept the extension regardless of that setting (as long as support is enabled at compile-time).

mpg commented:
@hanno-arm You left comments on an earlier version, can you please re-review?

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
 /**
  * \brief          Set the maximum fragment length to emit and/or negotiate
- *                 (Default: the smaller of MBEDTLS_SSL_IN_CONTENT_LEN and
+ *                 (Typical: the smaller of MBEDTLS_SSL_IN_CONTENT_LEN and
```
Minor: Please use `#` to have Doxygen generate a link to the documentation of `MBEDTLS_SSL_IN_CONTENT_LEN`.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  * \brief          Set the maximum fragment length to emit and/or negotiate
- *                 (Default: the smaller of MBEDTLS_SSL_IN_CONTENT_LEN and
+ *                 (Typical: the smaller of MBEDTLS_SSL_IN_CONTENT_LEN and
  *                 MBEDTLS_SSL_OUT_CONTENT_LEN, usually 2^14 bytes)
```
Same here.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
  *                 MBEDTLS_SSL_OUT_CONTENT_LEN, usually 2^14 bytes)
  *                 (Server: set maximum fragment length to emit,
- *                 usually negotiated by the client during handshake
+ *                 usually negotiated by the client during handshake)
```
Minor and pre-existing: I'm not a native, but `negotiated by [single entity]` doesn't sound right to me. 

I'd prefer to keep this overview section short, and provide details further below. What about:
```
   (Server: set the maximum fragment length to emit)
   (Client: set the maximum fragment length to emit and to communicate to the server)

\note On the server, the maximum fragment length extension doesn't provide a means to communicate
      the server's message size constraints to the client. If you're running a constrained server,
      you must ensure out of band that the client will not send messages larger than what the
      server can handle.
      On the client, the client will communicate its limitation to the server.
```

Feel free to elaborate on this, potentially drawing from https://tls.mbed.org/kb/how-to/controlling_package_size which explains everything in much more detail. Maybe we could even link the KB article here?

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2533,13 +2533,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
 #if defined(MBEDTLS_SSL_MAX_FRAGMENT_LENGTH)
 /**
  * \brief          Set the maximum fragment length to emit and/or negotiate
```
Minor: This is lacking a full stop.

k-stachowiak commented:
Thank you all for further comments. I tried addressing all of them.
@hanno-arm I made an attempt of restructuring the initial part of the comment, but, unfortunately I didn't feel that I can improve what is already there. Also, I would rather avoid referring from the code to external documentation due to a maintenance costs. So, as I think I cannot address that suggestion correctly, I decided I should not do it.

AndrzejKurek commented:
@mpg or @hanno-arm - please re-review this PR.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2532,14 +2532,19 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
- *                 (Default: the smaller of MBEDTLS_SSL_IN_CONTENT_LEN and
- *                 MBEDTLS_SSL_OUT_CONTENT_LEN, usually 2^14 bytes)
+ * \brief          Set the maximum fragment length to emit and/or negotiate.
+ *                 (Typical: the smaller of #MBEDTLS_SSL_IN_CONTENT_LEN and
```
It might be worth adding `(Default: #MBEDTLS_SSL_MAX_FRAG_LEN_NONE)` here.

k-stachowiak commented:
@hanno-arm Thanks for your review. I have address your comment; could you please re-review the PR?
@AndrzejKurek Could you please re-review the PR as well?

k-stachowiak commented:
@hanno-arm @AndrzejKurek As the backports have been approved, I added a change log entry here, which forced me to rebase this PR. Could you please re-review it?
